File: 210. Course Schedule II/210. Course Schedule II.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

class Solution {

  Map<Integer, List<Integer>> adjList;
  List<Integer> ansArrayList;
  Set<Integer> visiting;
  Set<Integer> visited;

  public int[] findOrder(int numCourses, int[][] prerequisites) {
    adjList = new HashMap<>();
    for (int[] pair : prerequisites) {
      adjList.computeIfAbsent(pair[0], k -> new ArrayList<>()).add(pair[1]);
    }

    ansArrayList = new ArrayList<>();

    visiting = new HashSet<>();
    visited = new HashSet<>();

    for (int course = 0; course < numCourses; course++) {
      if (!dfs(course)) {
        return new int[0];
      }
    }

    int[] ans = new int[numCourses];
    for (int i = 0; i < numCourses; i++) {
      ans[i] = ansArrayList.get(i);
    }
    return ans;
  }

  private boolean dfs(int currentCourse) {
    if (visiting.contains(currentCourse)) {
      return false;
    }

    if (visited.contains(currentCourse)) {
      return true;
    }

    visiting.add(currentCourse);

    if (adjList.containsKey(currentCourse)) {
      for (int course : adjList.get(currentCourse)) {
        if (!dfs(course)) {
          return false;
        }
      }
    }

    visiting.remove(currentCourse);
    visited.add(currentCourse);
    ansArrayList.add(currentCourse);
    return true;
  }
}

File: 210. Course Schedule II.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adj_list = defaultdict(list)
        for course, prereq in prerequisites:
            adj_list[course].append(prereq)

        ans = []
        visiting, visited = set(), set()

        def dfs(c):
            if c in visiting:
                return False

            if c in visited:
                return True

            visiting.add(c)

            for prereq in adj_list[c]:
                if not dfs(prereq):
                    return False

            visiting.remove(c)
            visited.add(c)
            ans.append(c)

            return True

        for c in range(numCourses):
            if c not in visited:
                if not dfs(c):
                    return []

        return ans
